src/main/java/com/google/sps/servlets/UserDataServlet.java

```diff
@@ -15,18 +15,138 @@
+      userDataBuilder.put(USER_EMAIL_PROPERTY, userEntity.getProperty(USER_EMAIL_PROPERTY));
+      userDataBuilder.put(USER_FRIENDS_LIST_PROPERTY, (ArrayList<String>) userEntity.getProperty(USER_FRIENDS_LIST_PROPERTY));
+      userDataBuilder.put(USER_ID_PROPERTY, userEntity.getProperty(USER_ID_PROPERTY));
+      userDataBuilder.put(USER_NAME_PROPERTY, userEntity.getProperty(USER_NAME_PROPERTY));
+    }
+
+    // Send the user's json data as the response
+    response.setContentType("application/json");
+    response.getWriter().print(gson.toJson(userDataBuilder.build()));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get the user's info to either update or create
+    String userId = getStringParameter(request, USER_ID_PROPERTY, DEFAULT_STRING);
+    String userName = getStringParameter(request, USER_NAME_PROPERTY, DEFAULT_STRING);
+    String userEmail = getStringParameter(request, USER_EMAIL_PROPERTY, DEFAULT_STRING);
+    String userBio = getStringParameter(request, USER_BIO_PROPERTY, DEFAULT_STRING);
+    String[] friends = getStringArrayParameter(request, USER_FRIENDS_LIST_PROPERTY, new String[]{});
+
+    // Check if a user entity with userId already exists
+    Entity userEntity = datastore.prepare(new Query(userId)).asSingleEntity();
+    if (userEntity == null) {
+      // User entity needs to be created
+      userEntity = new Entity(userId);
+      userEntity.setProperty(USER_ID_PROPERTY, userId);
+      userEntity.setProperty(USER_NAME_PROPERTY, userName);
+      userEntity.setProperty(USER_EMAIL_PROPERTY, userEmail);
+      userEntity.setProperty(USER_BIO_PROPERTY, userBio);
+      userEntity.setProperty(USER_FRIENDS_LIST_PROPERTY, Arrays.asList(friends));
+    }
+    else {
+      // User entity needs to be updated
+      // Set the properties of the user entity without overriding any values with the default string
+      setPropertyIfNotDefault(userEntity, USER_ID_PROPERTY, userId, DEFAULT_STRING);
+      setPropertyIfNotDefault(userEntity, USER_NAME_PROPERTY, userName, DEFAULT_STRING);
+      setPropertyIfNotDefault(userEntity, USER_EMAIL_PROPERTY, userEmail, DEFAULT_STRING);
+      setPropertyIfNotDefault(userEntity, USER_BIO_PROPERTY, userBio, DEFAULT_STRING);
+
+      // If the friends-list property wasn't given, don't override the current friends list
+      if (friends.length != 0) {
+        userEntity.setProperty(USER_FRIENDS_LIST_PROPERTY, Arrays.asList(friends));
+      }
+    }
+    datastore.put(userEntity);
+
+    // Redirect to the profile page, and let the front-end know the current logged in user
+    response.sendRedirect("/profile.html?id=" + userId);
+  }
+
+  /** Wrap the doGet method for jUnit testing */
+  public void doGetWrapper(DatastoreService datastoreService, HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
New idea - create a package private constructor that takes in the `DatastoreService`. Carefully document that it is for testing purposes only. At that point, you don't need these wrappers and the `datastore` variable can be final (I think - maybe not which is fine too).

You would initialize your fake data store before the servlet in your set up function of the test and be on your way. You'd also be able to remove the member variable for the data store in your test class (you can even probably do it inline in your constructor call).

Reply:
See https://github.com/googleinterns/step48-2020/pull/13#issuecomment-656928605